src/js/common/parser.js

```diff
@@ -0,0 +1,223 @@
+/*
+
+  A parser that dereferences `$ref` links within a JSON Schema so that we have a representation of the schema with
+  full definitions of each property.  This allows us to use the path data we get from the validator and find an
+  appropriate description field to use instead of the raw output.  Say we have a schema like the following:
+
+    {
+        type: "object",
+        definitions: {
+            field1: {
+                description: "Field one must contain at least one uppercase character.",
+                type:        "string",
+                pattern:     "[A-Z]+"
+            }
+        },
+        properties: {
+            field1: { $ref: "#/definitions/field1" }
+        }
+    }
+
+  If the validator tells us that `.field1` is invalid, it will produce error data like the following:
+
+    {
+        fieldErrors: {
+            field1: ["should match pattern \"[A-Z]\""]
+        }
+    }
+
+  The parser allows us to replace this with the description from the JSON Schema, as in:
+
+     {
+         fieldErrors: {
+             field1: ["Field one must contain at least one uppercase character."]
+         }
+     }
+
+  This is clearer to read in even the simplest case, and is especially useful when using `allOf` and `anyOf` to combine
+  rules.  Even a simple validation error for a field that uses `allOf` and `anyOf` would return at least two warnings
+  from the validator, one of which simply says how many rules failed, or that one of them must match.
+
+  If the desired schema property lacks a description, the raw message from the validator will be used as a fallback.
+
+  To use this component, you will need to instantiate it and make it aware of your schemas.  These will be dereferenced
+  and cached as they are added.  You can then use the `lookupField` method to retrieve any property of the field from
+  the schema definition, as in:
+
+    parser.lookupField("user", ".password.description")
+
+  This component uses [json-schema-ref-parser](https://github.com/BigstickCarpet/json-schema-ref-parser).  We configure
+  it using `options.parserOptions` (empty by default).  Read the library's documentation for details of what options
+  are available.
+
+  Note that this library expects to be able to resolve external references relative to the first argument passed to its
+  `dereference` function, and it expects to load and cache the references itself as needed.  Since we are receiving keys
+  from the `validator`, we need to some how convert these to URLs.  We do this using `fluid.stringTemplate` to combine
+  `options.schemaDir` with the individual key, which means:
+
+  1.  If `options.schemaDir` is a directory location, a full path will be produced.
+  2.  If `options.schemaDir` is a URL, a full URL will be produced.
+  3.  If `options.schemaDir` is empty, paths will be resolved relative to the working directory.
+
+  If you are using this with the `validator`, it takes care of that bit of wiring for you.
```

Comment:
What does the configuration look like, if I am using this together with the validator?